_posts/2020-02-03-apache-beam-how-beam-runs-on-top-of-flink.md

```diff
@@ -0,0 +1,157 @@
+1. A common pipeline format across all the languages: The Runner API
+2. A common interface during execution for the communication between the Runner and the code written in any language: The Fn API
+
+The Runner API provides a universal representation of the pipeline as Protobuf which contains the transforms, types, and the user code. Protobuf was chosen as the format because every language has libraries available for it. Similarly, for the execution part, Beam introduced the Fn API interface to handle the communication between the Runner/execution engine and the user code that may be written in a different language and executes in a different process. Fn API is pronounced "fun API", you may guess why.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/flink-runner-beam-language-portability.png" width="600px" alt="Language Portability in Apache Beam"/>
+</center>
+
+
+## How Are Beam Programs Translated In Language Portability?
+
+Users write their Beam pipelines in one language, but they may get executed in an environment based on a completely different language. How does that work? To explain that, let's follow the life cycle of a pipeline. Let's suppose we use the Python SDK to write the pipeline. Before we would submit the pipeline via the Job API to Beam's JobServer, Beam would convert it to the Runner API, the language-agnostic format we described before. The JobServer is also a Beam component that handles the staging of the required dependencies during execution. The JobServer will then kick-off the translation which is similar to the classic Runner. However, an important change is the so-called `ExecutableStage` transform. It is essentially a ParDo transform which we already know, but designed for holding language-dependent code. Beam tries to combine as many of these transforms into one "executable stage". The result again is a Flink program which is then sent to the Flink cluster and executed there. The major difference compared to the classic Runner is that during execution we will start _environments_ to execute the aforementioned _ExecutableStages_. The following environments are available:
+
+- Docker-based (the default)
+- Process-based (a simple process is started)
+- Externally-provided (K8s or other schedulers)
+- Embedded (intended for testing and only works with Java)
+
+Environments hold the _SDK Harness_ which is the code that handles the execution and the communication with the Runner over the Fn API. For example, when Flink executes Python code, it sends the data to the Python environment containing the Python SDK Harness. Sending data to an external process involves a minor overhead which we have measured to be 5-10% slower than the classic Java pipelines. However, Beam uses fusion of transforms to execute as many transforms as possible in the same environment which share the same input or output. That's why, In real-world scenarios the overhead could be much lower.
```

Comment:
```suggestion
Environments hold the _SDK Harness_ which is the code that handles the execution and the communication with the Runner over the Fn API. For example, when Flink executes Python code, it sends the data to the Python environment containing the Python SDK Harness. Sending data to an external process involves a minor overhead which we have measured to be 5-10% slower than the classic Java pipelines. However, Beam uses a fusion of transforms to execute as many transforms as possible, in the same environment, which share the same input or output. That's why in real-world scenarios the overhead could be much lower.
```

Reply:
This is an outdated diff. I cannot incorporate this suggestion from Github. Please comment only on the current version.